### PR TITLE
Making TagBuilder implement IHtmlContent and removing ToHtmlContent.

### PR DESCRIPTION
**Pull request by sornaks**
- Making `TagRenderMode` a property in `TagBuilder`.
- Modifying places where `TagBuilder` is used to suit the new model.
- This reduces space occupied by a normal application by 11.8%. (Tested over 1000 requests with 20 in parallel)

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs [hunk 1]
```diff
@@ -236,25 +236,27 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
                 }
             }
 
-            var tagBuilder = Generator.GenerateCheckBox(
+            var checkBoxTag = Generator.GenerateCheckBox(
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: Curious: Why did you change from reusing the one `tagBuilder` variable?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs [hunk 2]
```diff
@@ -237,23 +234,16 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
...
-                        divTag.AddCssClass("display-label");
-                        content.AppendLine(divTag.ToHtmlContent(TagRenderMode.Normal));
-
-                        // Reset divTag for reuse.
-                        divTag.Attributes.Clear();
-                    }
-
-                    divTag.AddCssClass("display-field");
-                    content.Append(divTag.ToHtmlContent(TagRenderMode.StartTag));
+                        var labelTag = new TagBuilder("div");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: `divTag`; it's not a `<label>`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs [hunk 3]
```diff
@@ -237,23 +234,16 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
...
                 {
                     var label = propertyMetadata.GetDisplayName();
                     if (!string.IsNullOrEmpty(label))
                     {
-                        divTag.SetInnerText(label);
-                        divTag.AddCssClass("display-label");
-                        content.AppendLine(divTag.ToHtmlContent(TagRenderMode.Normal));
-
-                        // Reset divTag for reuse.
-                        divTag.Attributes.Clear();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: Why not reuse the same `TagBuilder` instance?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs [hunk 4]
```diff
@@ -266,11 +256,16 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
                     readOnly: true,
                     additionalViewData: null);
 
-                content.Append(templateBuilder.Build());
-
                 if (!propertyMetadata.HideSurroundingHtml)
                 {
-                    content.AppendLine(divTag.ToHtmlContent(TagRenderMode.EndTag));
+                    var containerDivTag = new TagBuilder("div");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: `divTag` or perhaps `valueDivTag`; it's no more a "container" than any other `<div>` element.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs [hunk 5]
```diff
@@ -266,11 +256,16 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
                     readOnly: true,
                     additionalViewData: null);
 
-                content.Append(templateBuilder.Build());
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: stick `templateBuilder.Build()` into a variable to make it blindingly obvious it's called without regard to `HideSurroundingHtml`.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs [hunk 6]
```diff
@@ -266,16 +264,11 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
...
                     {
-                        divTag.AddCssClass("editor-label");
-                        divTag.InnerHtml = label; // already escaped
-                        content.AppendLine(divTag.ToHtmlContent(TagRenderMode.Normal));
-
-                        // Reset divTag for reuse.
-                        divTag.Attributes.Clear();
+                        var labelDivTag = new TagBuilder("div");
+                        labelDivTag.AddCssClass("editor-label");
+                        labelDivTag.InnerHtml = label; // already escaped
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: Remove `// already escaped` comment since `htmlHelper.Label()` returns an `IHtmlContent`, not a `string`.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs [hunk 7]
```diff
@@ -266,16 +264,11 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
                         htmlAttributes: null);
                     if (!string.IsNullOrEmpty(label.ToString()))
                     {
-                        divTag.AddCssClass("editor-label");
-                        divTag.InnerHtml = label; // already escaped
-                        content.AppendLine(divTag.ToHtmlContent(TagRenderMode.Normal));
-
-                        // Reset divTag for reuse.
-                        divTag.Attributes.Clear();
+                        var labelDivTag = new TagBuilder("div");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: Rinse and repeat comments from `DefaultDisplayTemplates`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs [hunk 8]
```diff
@@ -288,18 +281,26 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
...
-                content.Append(templateBuilder.Build());
-
                 if (!propertyMetadata.HideSurroundingHtml)
                 {
-                    content.Append(" ");
-                    content.Append(htmlHelper.ValidationMessage(
+                    var divTag = new TagBuilder("div");
+                    divTag.AddCssClass("editor-field");
+
+                    var innerContent = new BufferedHtmlContent();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: Not sure what refactoring to avoid `TagRenderModel.StartTag` and `EndTag` buys us. This extra buffer and the new lack of alignment between the code and generated HTML are downsides.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: Oh, I see. Can't put the same `TagBuilder` into the final content. Might do with a couple of helper methods to restore alignment between code and generated HTML.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs [hunk 9]
```diff
@@ -1302,12 +1302,26 @@ private IHtmlContent GenerateGroupsAndOptions(string optionLabel, IEnumerable<Se
...
-                TagBuilder groupBuilder = null;
                 if (optGroup != null)
                 {
-                    groupBuilder = new TagBuilder("optgroup");
+                    optGroupContent = new BufferedHtmlContent();
+                    optGroupContent.Append(Environment.NewLine);
+                }
+                else
+                {
+                    optGroupContent = listItemBuilder;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs]: This and [lines 1335,1336](https://github.com/aspnet/Mvc/pull/2952/files#diff-0e83439eeed8ffbde4eceb0db1f6e541R1336) mean the `listItemBuilder` is placed in the `listItemBuilder` content. How does this not stack dive?

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs [hunk 10]
```diff
@@ -236,25 +236,27 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
...
-            if (tagBuilder != null)
+            if (checkBoxTag != null)
             {
                 // Do not generate current element's attributes or tags. Instead put both <input type="checkbox"/> and
                 // <input type="hidden"/> into the output's Content.
                 output.Attributes.Clear();
                 output.TagName = null;
 
-                output.Content.Append(tagBuilder.ToHtmlContent(TagRenderMode.SelfClosing));
+                checkBoxTag.TagRenderMode = TagRenderMode.SelfClosing;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: Instead set `TagRenderMode` when creating the `TagBuilder` instance. Current separation between `TagBuilder` content and whether or not the element should be self-closing is artificial, just a result of the previous APIs. Might even be worth adding the mode or a `bool isSelfClosing` parameter to a `TagBuilder` constructor overload, saving the property setter for the (rare) `StartTag` / `EndTag` special cases.

Here for example `DefaultHtmlGenerator` knows the `<input>` tag is self-closing and this code shouldn't care.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs (hunk 2 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: But it *is* a tag containing a label. Calling it `divTag` would honestly be more akin to Hungarian notation. Identifiers should convey *semantics*, not *types*.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: Well, most `TagBuilder` variables in product code are named `tagBuilder` which truly conveys the _type_. Even that works better here than `labelTag` because the semantics of the generated `<div>` are not the same as the contained `<label>` element.

BTW we also name `TagBuilder` variables after the generated element many other places, starting probably in MVC 1.0. The reasoning might be along the lines "The _semantics_ of this `TagBuilder` are that it will generate a `<div>` element." Are you suggesting we make a general change throughout MVC?

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs (hunk 10 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: I don't think combining a `bool` and `TagBuilderMode` makes things any simpler. I'd like to stick to having one way to set the mode. Easier to document, easier to understand.

But one thing that I think would be nice is if it would be possible to set the mode in the generator as opposed to out here.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: Works for me. Hence my "Might even" prefix to the "any simpler" sentence :smiley_cat:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs (hunk 2 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: Yes, but those are almost always wrong. Just like calling all StringBuilders `sb`.

In some cases there's nothing more to say about the tag builder other than `divBuilder` so it's not necessarily wrong. In this case it's not just "any" div builder - it's for one that contains a label.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: @Eilon at least for me `labelTag` is confusing because this builder will not generate a `<label>`. That name doesn't say anything about what the builder will do, just what the generated body will be.

Would `labelDivBuilder` (or perhaps `nameDivBuilder` since this is a display template) work here and `valueDivBuilder` just below? If not what names do you suggest?

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultDisplayTemplates.cs]: I don't think it's worth changing.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs (hunk 8 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: I don't understand. Where has the alignment changed?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs (hunk 9 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs]: No. <a href="https://github.com/aspnet/Mvc/pull/2952/files#diff-0e83439eeed8ffbde4eceb0db1f6e541R1336">Lines 1335,1336</a> are executed only if `optGroup` is <i>not</i> null. So `listItemBuilder` will never go into the content of `listItemBuilder`.

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs (hunk 10 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: I think the Generator should not set the <i>Render</i> mode. I should be able to generate a tag and choose to render it as self-closing or as a normal tag with start and end tags (we do this in a few places in our code).

**Comment by sornaks** [comment]
Updated..

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs (hunk 10 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: Ah then nevermind.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: Sorry, I'm not understanding this "Won't fix". The HTML specification makes it clear that end tags are illegal for void elements e.g. `<input>`. The HTML generator deals with such HTML semantics. Why instead spread the information around?

Separately I mentioned leaving handling of start and end tags alone in my original comment. That's a separate concern from whether the end tag is valid at all.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs (hunk 9 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs]: This is a relatively short `foreach` block but it remains difficult to grok. Suggest repeating the inner `item in group` loop within a single `if (optGroup == null) {} else {}` statement.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs [hunk 11]
```diff
@@ -288,18 +281,27 @@ public static IHtmlContent ObjectTemplate(IHtmlHelper htmlHelper)
                     readOnly: false,
                     additionalViewData: null);
 
-                content.Append(templateBuilder.Build());
-
+                var templateBuilderResult = templateBuilder.Build();
                 if (!propertyMetadata.HideSurroundingHtml)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: The new structure may be _too_ similar to the previous code. Suggest seeing if the loop is more clear with a single `propertyMetadata.HideSurroundingHtml` check, similar to my recent `DefaultHtmlGenerator` comment. Could do the same thing in `DefaultDisplayTemplates` but there's less inside the conditional blocks there.

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs (hunk 10 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: <a href="https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs#L170">This</a> is the place I am referring to. Not all generated tags are self-closing except checkbox.

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: What? Definitely not just `<input type="checkbox" />` alone. All void elements (``, `<input>`, `<link>`, and more) must be written out as self-closing (or start tag only once @NTaylorMullen gets void element support in). Fortunately the default rendering mode works fine for most of what `DefaultHtmlGenerator` generates.

**Comment by sornaks** [comment]
Updated..

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs [hunk 12]
```diff
@@ -1302,12 +1304,17 @@ private IHtmlContent GenerateGroupsAndOptions(string optionLabel, IEnumerable<Se
             foreach (var group in groupedSelectList)
             {
                 var optGroup = group.First().Group;
-
-                // Wrap if requested.
-                TagBuilder groupBuilder = null;
                 if (optGroup != null)
                 {
-                    groupBuilder = new TagBuilder("optgroup");
+                    var optGroupContent = new BufferedHtmlContent();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs]: Suggest moving this down closer to the first use (`groupBuilder.InnerHtml = optGroupContent;`)

Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs]: This is used in <a href="https://github.com/aspnet/Mvc/pull/2952/files#diff-0e83439eeed8ffbde4eceb0db1f6e541R1314">line 1314</a>.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs (hunk 11 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultEditorTemplates.cs]: thx; looks good

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs (hunk 12 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/Html/DefaultHtmlGenerator.cs]: Agreed, I meant to move that `foreach` as well.

**Comment by dougbu** [comment]
:shipit:

**Comment by sornaks** [comment]
Checked in. Thanks! https://github.com/aspnet/Mvc/commit/f2540f9ba19937f79c1a780ef12dd9a9d5cd84a6
